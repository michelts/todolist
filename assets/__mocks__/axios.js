export default {
  get: jest.fn(),
  post: jest.fn(),
  put: jest.fn(),
};
