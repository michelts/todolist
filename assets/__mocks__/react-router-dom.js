export const useHistory = jest.fn();

export const Link = () => null;
