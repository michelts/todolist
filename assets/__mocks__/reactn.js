export const useGlobal = jest.fn();
