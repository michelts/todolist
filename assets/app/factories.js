import Factory from 'rosie';

const UserFactory = new Factory()
  .sequence('id', (index) => index + 1)
  .sequence('username', (index) => `username_${index}`)
  .sequence('name', (index) => `Name-${index}`);

export {
  UserFactory,
};
