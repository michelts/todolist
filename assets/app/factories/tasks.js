import { Factory } from 'rosie';

const TaskFactory = new Factory()
  .sequence('id', (index) => index + 1)
  .sequence('description', (index) => `Task ${index}`)
  .attr('priority', 1)
  .attr('due_date', '2020-05-22')
  .attr('is_completed', false);

export {
  TaskFactory,
}
