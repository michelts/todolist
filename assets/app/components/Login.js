import React from 'react';
import axios from 'axios';
import { useHistory } from 'react-router-dom';
import { Formik, Field } from 'formik';
import * as Yup from 'yup';

export const LoginSchema = Yup.object().shape({
  username: Yup.string().email().required(),
  password: Yup.string().required(),
});

const Login = () => {
  const history = useHistory();
  const [loginFailed, setLoginFailed] = React.useState(false);

  React.useEffect(() => {
    axios.get('/api/v1/users/current/')
      .then((user) => history.push('/tasks'))
      .catch(() => {});
  }, []);

  const initialValues = {
    username: '',
    password: '',
  };

  const handleSubmit = (values) => {
    axios.post('/api/v1/users/login/', values)
      .then((user) => history.push('/tasks'))
      .catch(() => setLoginFailed(true));
  };

  return (
    <Formik
      initialValues={initialValues}
      onSubmit={handleSubmit}
    >
      {loginFailed && (
        <p>Login failed! Check your credentials!</p>
      )}

      <Field type="text" name="username" label="Username" />

      <Field type="password" name="password" label="Password" />

      <button type="submit">Submit</button>
    </Formik>
  );
};

export default Login;
