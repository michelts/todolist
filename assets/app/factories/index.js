import { UserFactory } from './user';
import { TaskFactory, BlankTaskFactory } from './tasks';

export {
  UserFactory,
  TaskFactory,
  BlankTaskFactory,
};
