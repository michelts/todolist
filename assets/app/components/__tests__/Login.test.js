import React from 'react';
import { useGlobal } from 'reactn';
import { shallow } from 'enzyme';
import axios from 'axios';
import { useHistory } from 'react-router-dom';
import { Formik } from 'formik';
import { UserFactory } from '../../factories';
import Login from '../Login';

const getWrapper = () => {
  const history = {
    push: jest.fn(),
  };
  useHistory.mockReturnValue(history);
  const wrapper = shallow(<Login />);
  return { wrapper, history };
};

const formikChild = (wrapper) => {
  const handleSubmit = jest.fn().mockName('formikSubmit');
  return wrapper.find(Formik).renderProp('children')({ handleSubmit });
};

describe('Login component', () => {
  let setUser;

  beforeEach(() => {
    setUser = jest.fn();
    jest.spyOn(React, 'useEffect').mockImplementation((effect) => effect());
  });

  it('should render a loading component while the current user checking is being processed', async () => {
    useGlobal.mockReturnValue([undefined, setUser]);
    axios.get.mockRejectedValue({});

    const { wrapper } = getWrapper();
    expect(wrapper).toMatchSnapshot();
  });

  it('should set global user and redirect to the task list if the user is already logged in, but only when mounted', async () => {
    useGlobal.mockReturnValue([undefined, setUser]);

    const user = UserFactory.build();
    axios.get.mockResolvedValue({ data: user });

    const { history } = getWrapper();
    expect(axios.get).toHaveBeenCalledWith('/api/v1/users/current/');

    await axios.get;
    expect(history.push).toHaveBeenCalledWith('/tasks');
    expect(setUser).toHaveBeenCalledWith(user);

    // Enzyme doesn't allow to test effect update yet so, to ensure we are
    // acting only on mount, we test the signature passed to useEffect.
    expect(React.useEffect).toHaveBeenCalledWith(expect.anything(), []);
  });

  it('should setUser to null if current user is not logged in, when mounted', async (done) => {
    useGlobal.mockReturnValue([undefined, setUser]);
    axios.get.mockRejectedValue({});

    const { history } = getWrapper();
    expect(axios.get).toHaveBeenCalledWith('/api/v1/users/current/');

    process.nextTick(() => {
      expect(history.push).not.toHaveBeenCalled();
      expect(setUser).toHaveBeenCalledWith(null);
      done();
    });
  });

  it('should render a login form if global user is null (e.g. not authenticated)', async () => {
    useGlobal.mockReturnValue([null, setUser]);
    axios.get.mockRejectedValue({});

    const { wrapper } = getWrapper();
    await axios.get;

    expect(wrapper).toMatchSnapshot('Formik');
    expect(formikChild(wrapper)).toMatchSnapshot('Formik children');
  });

  it('should authenticate the user and redirect him to the task list when form is submitted successfully', async () => {
    useGlobal.mockReturnValue([null, setUser]);
    axios.get.mockRejectedValue({});

    const loggedInUser = UserFactory.build();
    axios.post.mockResolvedValue({ data: loggedInUser });

    const { wrapper, history } = getWrapper();
    await axios.get;

    const payload = { username: 'foo', password: 'bar' };
    wrapper.find(Formik).simulate('submit', payload);
    expect(axios.post).toHaveBeenCalledWith('/api/v1/users/login/', payload);

    await axios.post;
    expect(history.push).toHaveBeenCalledWith('/tasks');
    expect(setUser).toHaveBeenCalledWith(loggedInUser);
  });

  it('should indicate authenticated failed if form submission is not successful', async (done) => {
    useGlobal.mockReturnValue([null, setUser]);
    axios.get.mockRejectedValue({});
    axios.post.mockRejectedValue({});

    const { wrapper, history } = getWrapper();
    await axios.get;

    const payload = { username: 'foo', password: 'bar' };
    wrapper.find(Formik).simulate('submit', payload);
    expect(axios.post).toHaveBeenCalledWith('/api/v1/users/login/', payload);

    await axios.post;
    expect(history.push).not.toHaveBeenCalled();

    process.nextTick(() => {
      expect(wrapper).toMatchSnapshot('Error msg and Formik');
      expect(formikChild(wrapper)).toMatchSnapshot('Formik children');
      done();
    });
  });
});
