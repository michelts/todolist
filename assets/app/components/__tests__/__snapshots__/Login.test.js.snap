// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component should indicate authenticated failed if form submission is not successful 1`] = `
<Formik
  initialValues={
    Object {
      "password": "",
      "username": "",
    }
  }
  onSubmit={[Function]}
>
  <p>
    Login failed! Check your credentials!
  </p>
  <Field
    label="Username"
    name="username"
    type="text"
  />
  <Field
    label="Password"
    name="password"
    type="password"
  />
  <button
    type="submit"
  >
    Submit
  </button>
</Formik>
`;

exports[`Login component should render a login form if user is not authenticated 1`] = `
<Formik
  initialValues={
    Object {
      "password": "",
      "username": "",
    }
  }
  onSubmit={[Function]}
>
  <Field
    label="Username"
    name="username"
    type="text"
  />
  <Field
    label="Password"
    name="password"
    type="password"
  />
  <button
    type="submit"
  >
    Submit
  </button>
</Formik>
`;
